hare4/hare.go

```diff
@@ -283,6 +308,135 @@ func (h *Hare) Running() int {
 	return len(h.sessions)
 }
 
+// fetchFull will fetch the full list of proposal IDs from the provided peer.
+func (h *Hare) fetchFull(ctx context.Context, peer p2p.Peer, msgId types.Hash32) (
+	[]types.ProposalID, error,
+) {
+	ctx, cancel := context.WithTimeout(ctx, fetchFullTimeout)
+	defer cancel()
+
+	requestCompactCounter.Inc()
+	req := &CompactIdRequest{MsgId: msgId}
+	reqBytes := codec.MustEncode(req)
+	resp := &CompactIdResponse{}
+	cb := func(ctx context.Context, rw io.ReadWriter) error {
+		respLen, _, err := codec.DecodeLen(rw)
+		if err != nil {
+			return fmt.Errorf("decode length: %w", err)
+		}
+		if respLen >= MAX_EXCHANGE_SIZE {
+			return errResponseTooBig
+		}
+		buff := make([]byte, respLen)
+		_, err = io.ReadFull(rw, buff)
+		if err != nil {
+			return fmt.Errorf("read response buffer: %w", err)
+		}
+		err = codec.Decode(buff, resp)
+		if err != nil {
+			return fmt.Errorf("decode response: %w", err)
+		}
+		return nil
+	}
+
+	err := h.p2p.StreamRequest(ctx, peer, reqBytes, cb)
+	if err != nil {
+		requestCompactErrorCounter.Inc()
+		return nil, fmt.Errorf("stream request: %w", err)
+	}
+
+	h.tracer.OnCompactIdResponse(resp)
+
+	return resp.Ids, nil
+}
+
+func (h *Hare) handleProposalsStream(ctx context.Context, msg []byte, s io.ReadWriter) error {
+	requestCompactHandlerCounter.Inc()
+	compactProps := &CompactIdRequest{}
+	if err := codec.Decode(msg, compactProps); err != nil {
+		malformedError.Inc()
+		return fmt.Errorf("%w: decoding error %s", pubsub.ErrValidationReject, err.Error())
+	}
+	h.tracer.OnCompactIdRequest(compactProps)
+	h.mu.Lock()
+	m, ok := h.messageCache[compactProps.MsgId]
+	h.mu.Unlock()
+	if !ok {
+		messageCacheMiss.Inc()
+		return fmt.Errorf("message %s: cache miss", compactProps.MsgId)
+	}
+	resp := &CompactIdResponse{Ids: m.Body.Value.Proposals}
+	respBytes := codec.MustEncode(resp)
+	if _, err := codec.EncodeLen(s, uint32(len(respBytes))); err != nil {
+		return fmt.Errorf("encode length: %w", err)
+	}
+
+	if _, err := s.Write(respBytes); err != nil {
+		return fmt.Errorf("write response: %w", err)
+	}
+
+	return nil
+}
+
+// reconstructProposals tries to reconstruct the full list of proposals from a peer based on a delivered
+// set of compact IDs.
+func (h *Hare) reconstructProposals(ctx context.Context, peer p2p.Peer, msgId types.Hash32, msg *Message) error {
+	proposals := h.proposals.GetForLayer(msg.Layer)
+	if len(proposals) == 0 {
+		return errNoLayerProposals
+	}
+	compacted := h.compactProposals(h.compactFn, msg.Layer, proposals)
+	proposalIds := make([]proposalTuple, len(proposals))
+	for i := range proposals {
+		proposalIds[i] = proposalTuple{id: proposals[i].ID(), compact: compacted[i]}
+	}
+	slices.SortFunc(proposalIds, sortProposalsTuple)
```

Comment:
Why is it sorted?

Reply:
It is sorted on both sides actually - both on the original proposer:
```go
func (p *protocol) OnInitial(proposals []types.ProposalID) {
	slices.SortFunc(proposals, sortProposalIds)
	p.mu.Lock()
	defer p.mu.Unlock()
	p.initial = proposals
}
```
And above ^^.
The assumption is that given both sets of proposals are visible from both proposer and the verifier of the message, given there are duplicate prefixes of certain messages, that they are processed in order, this assures that the verifier can verify the message in the order it was created on the proposer side.